### PR TITLE
CRM-307: Schema change for civicrm_ch_contribution_batch table

### PR DESCRIPTION
**Pull request by monishdeb**
I have added the upgrade and install code to change the schema of ```civicrm_ch_contribution_batch``` table where : 
1. ```ch_fund``` length is increased to 255 from 10
2. ```fund length``` is increased to 11 from 4
3. ```contribution_id``` is increased to 11 from 10

### CODE REVIEW

File: sql/auto_install.sql [hunk 1]
```diff
@@ -96,8 +96,8 @@ CREATE TABLE `civicrm_option_value_ch` (
 )    ;
 
 CREATE TABLE IF NOT EXISTS `civicrm_ch_contribution_batch` (
-   `id` INT NOT NULL AUTO_INCREMENT , `ch_fund` VARCHAR(10) NOT NULL ,
-   `fund` INT(4) NOT NULL ,
-   `contribution_id` INT(10) NOT NULL ,
+   `id` INT NOT NULL AUTO_INCREMENT , `ch_fund` varchar(255) NOT NULL ,
```
Comment (JoeMurray) [inline sql/auto_install.sql]: Please put 'ch_fund' onto a different line.

File: CRM/Chfunds/Upgrader.php [hunk 2]
```diff
@@ -26,6 +26,18 @@ public function upgrade_1200() {
     return TRUE;
   }
 
+  public function upgrade_1300() {
+    $this->ctx->log->info('Applying update 1.3');
+    $sql = "
+    ALTER TABLE `civicrm_ch_contribution_batch`
+      CHANGE `ch_fund` `ch_fund` VARCHAR(255) CHARACTER SET utf8 COLLATE utf8_general_ci NOT NULL,
```
Comment (JoeMurray) [inline CRM/Chfunds/Upgrader.php]: Collation should be uft8_unicode_ci.

File: sql/auto_install.sql (hunk 1 above)
Comment (JoeMurray) [reply sql/auto_install.sql]: We should ensure collation is same as rest of db, which is utf8_unicode_ci.

File: sql/auto_install.sql [hunk 3]
```diff
@@ -96,8 +96,9 @@ CREATE TABLE `civicrm_option_value_ch` (
 )    ;
 
 CREATE TABLE IF NOT EXISTS `civicrm_ch_contribution_batch` (
-   `id` INT NOT NULL AUTO_INCREMENT , `ch_fund` VARCHAR(10) NOT NULL ,
-   `fund` INT(4) NOT NULL ,
-   `contribution_id` INT(10) NOT NULL ,
+   `id` INT NOT NULL AUTO_INCREMENT ,
+   `ch_fund` varchar(255) NOT NULL ,
```
Comment (JoeMurray) [inline sql/auto_install.sql]: pls add collation utf8_unicode_ci

**Comment by seamuslee001** [comment]
@monishdeb this looks good to me but can you please confirm where the spec is for the size needed. I just see in the description of the ticket to increase the size but just wanting to ensure we aren't going too big or too small

**Comment by monishdeb** [comment]
@seamuslee001 back then the CHFund hit an issue for which we decided to process the contributions in the backend schedule job and here's the discussion in MM https://hub.symbiotic.coop/coopsymbiotic/pl/uqx8kunacp8y3mc47qbkm7z4pr This spec doesn't specify the size of the batch table. Recently Irfan made an [request](https://hub.symbiotic.coop/coopsymbiotic/pl/xtpwoqksm7y9jpoqg4ca5rnmyc) to increase the ch_fund column length, and along with it, I have increased the size of contribution ID and fund ID which was too small for a large database.

**Review by seamuslee001 — COMMENTED** [review]
Changes look fine to me will try to run the upgrader shortly

**Comment by JoeMurray** [comment]
There was a failure when the batch update process ran as the keys didn't match. So we should look at code where these fields are being compared. I bet it was the varchar

**Comment by seamuslee001** [comment]
I tested and confirmed the upgrade worked successfully

```
mysql> show create table civicrm_ch_contribution_batch;
+-------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table                         | Create Table                                                                                                                                                                                                                                                                                      |
+-------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| civicrm_ch_contribution_batch | CREATE TABLE `civicrm_ch_contribution_batch` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `ch_fund` varchar(255) CHARACTER SET utf8 COLLATE utf8_unicode_ci NOT NULL,
  `fund` int(11) NOT NULL,
  `contribution_id` int(11) NOT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=latin1 |
+-------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

```

**Review by seamuslee001 — APPROVED** [review]

**Comment by seamuslee001** [comment]
This is fine to merge from my POV @JoeMurray @monishdeb

**Comment by seamuslee001** [comment]
@monishdeb just thinking through is there any reason not to change the ch_fund column to be 512 as the current option_value schema is

```
`value` varchar(512) COLLATE utf8_unicode_ci NOT NULL COMMENT 'The actual value stored (as a foreign key) in the data record. Functions which need lookup option_value.title should use civicrm_option_value.option_group_id plus civicrm_option_value.value as the key.',
```
The ints changes seem to be fine no matter what

**Comment by monishdeb** [comment]
@seamuslee001 in my opinion as ```value``` meant to store CH Fund value only which can't be that large. On the other hand core sometimes needs to store large string to which are usually serialized string for Mailing label_format, PDF format etc. reason why it needs huge length of 512 for optionvalue.

**Comment by seamuslee001** [comment]
@monishdeb i think 255 is a reasonable length, Just wanting to verbalise so we are clear on why the decision was made

**Pull request merged**